web_poet/overrides.py

```diff
@@ -0,0 +1,383 @@
+import importlib
+import importlib.util
+import warnings
+import pkgutil
+from collections import deque
+from dataclasses import dataclass, field
+from types import ModuleType
+from typing import Iterable, Optional, Union, List, Callable, Dict, Any
+
+from url_matcher import Patterns
+
+Strings = Union[str, Iterable[str]]
+
+
+@dataclass
+class OverrideRule:
+    """A single override rule that specifies when a Page Object should be used
+    instead of another.
+
+    This is instantiated when using the :func:`web_poet.handle_urls` decorator.
+    It's also being returned as a ``List[OverrideRule]`` when calling
+    :meth:`~.PageObjectRegistry.get_overrides`.
+
+    You can access any of its attributes:
+
+        * ``for_patterns: Patterns`` - contains the URL patterns associated
+          with this rule. You can read the API documentation of the
+          `url-matcher <https://url-matcher.readthedocs.io/>`_ package for more
+          information.
+        * ``use: Callable`` - the Page Object that will be used.
+        * ``instead_of: Callable`` - the Page Object that will be **replaced**.
+        * ``meta: Dict[str, Any] = field(default_factory=dict)`` - Any other
+          information you many want to store. This doesn't do anything for now
+          but may be useful for future API updates.
+    """
+
+    for_patterns: Patterns
+    use: Callable
+    instead_of: Callable
+    meta: Dict[str, Any] = field(default_factory=dict)
+
+    def __hash__(self):
+        # TODO: Remove this when the following has been implemented:
+        #   - https://github.com/zytedata/url-matcher/issues/3
+        pattern_hash = hash(
+            (
+                tuple(self.for_patterns.include),
+                tuple(self.for_patterns.exclude),
+                self.for_patterns.priority,
+            )
+        )
+        return hash((pattern_hash, self.use, self.instead_of))
```

Comment:
@kmike following our conversation in https://github.com/zytedata/url-matcher/pull/4, `OverrideRule` should be set to **frozen=True** as well.

Although this changes the use case I've written down on the docs for scenarios wherein:

- users are importing an external library of Page Objects
- manipulates the existing Registry rules from it _(e.g. improving rules, replacing POs, etc)_

I'll update the guide in the doc for this one. However, making `OverrideRule` frozen would need a new API to make it more easy to manipulate the rules.

I'll push a commit on this one soon.

Reply:
Fixed this in https://github.com/scrapinghub/web-poet/pull/16/commits/bf0b3e5e04b3bbb2a8e25f9118d271727fc4f3a4.